From 4bfe26268019f209f8c97b01361ac87fcd2aacf6 Mon Sep 17 00:00:00 2001
From: Mitch Curtis <mitch.curtis@qt.io>
Date: Fri, 5 Sep 2025 15:43:42 +0800
Subject: [PATCH 012/239] Fusion: give TextArea a background

This is the same background as TextField. This will cause the
implicitWidth for TextArea to be the same as TextField - i.e. a bit
wider - whereas before it was only as wide as the text + padding.

This fixes it not being visible, especially without placeholder text.

Fixes: QTBUG-139715
Pick-to: 6.9
Change-Id: Ia736c2e282497f0a40bf8726adcc2838082ea5b4
Reviewed-by: Oliver Eftevaag <oliver.eftevaag@qt.io>
(cherry picked from commit 006c4eedda9c42b4f47a0b9c5c11277dc341369b)
---
 src/quickcontrols/fusion/TextArea.qml         |  5 ++-
 src/quickcontrols/fusion/TextField.qml        | 27 ++------------
 src/quickcontrols/fusion/impl/CMakeLists.txt  |  1 +
 .../fusion/impl/TextFieldBackground.qml       | 35 +++++++++++++++++++
 4 files changed, 42 insertions(+), 26 deletions(-)
 create mode 100644 src/quickcontrols/fusion/impl/TextFieldBackground.qml

diff --git a/src/quickcontrols/fusion/TextArea.qml b/src/quickcontrols/fusion/TextArea.qml
index 5e1ecfa905..377c1c20b1 100644
--- a/src/quickcontrols/fusion/TextArea.qml
+++ b/src/quickcontrols/fusion/TextArea.qml
@@ -4,7 +4,6 @@
 import QtQuick
 import QtQuick.Templates as T
 import QtQuick.Controls.impl
-import QtQuick.Controls.Fusion
 import QtQuick.Controls.Fusion.impl
 
 T.TextArea {
@@ -40,4 +39,8 @@ T.TextArea {
         elide: Text.ElideRight
         renderType: control.renderType
     }
+
+    background: TextFieldBackground {
+        control: control
+    }
 }
diff --git a/src/quickcontrols/fusion/TextField.qml b/src/quickcontrols/fusion/TextField.qml
index b358c668d8..7f8775f8bb 100644
--- a/src/quickcontrols/fusion/TextField.qml
+++ b/src/quickcontrols/fusion/TextField.qml
@@ -40,30 +40,7 @@ T.TextField {
         renderType: control.renderType
     }
 
-    background: Rectangle {
-        implicitWidth: 120
-        implicitHeight: 24
-
-        radius: 2
-        color: control.palette.base
-        border.color: control.activeFocus ? Fusion.highlightedOutline(control.palette) : Fusion.outline(control.palette)
-
-        Rectangle {
-            x: 1; y: 1
-            width: parent.width - 2
-            height: parent.height - 2
-            color: "transparent"
-            border.color: Color.transparent(Fusion.highlightedOutline(control.palette), 40 / 255)
-            visible: control.activeFocus
-            radius: 1.7
-        }
-
-        Rectangle {
-            x: 2
-            y: 1
-            width: parent.width - 4
-            height: 1
-            color: Fusion.topShadow
-        }
+    background: TextFieldBackground {
+        control: control
     }
 }
diff --git a/src/quickcontrols/fusion/impl/CMakeLists.txt b/src/quickcontrols/fusion/impl/CMakeLists.txt
index 64ee4851fb..79080289be 100644
--- a/src/quickcontrols/fusion/impl/CMakeLists.txt
+++ b/src/quickcontrols/fusion/impl/CMakeLists.txt
@@ -12,6 +12,7 @@ set(qml_files
     "SliderGroove.qml"
     "SliderHandle.qml"
     "SwitchIndicator.qml"
+    "TextFieldBackground.qml"
 )
 
 qt_internal_add_qml_module(QuickControls2FusionStyleImpl
diff --git a/src/quickcontrols/fusion/impl/TextFieldBackground.qml b/src/quickcontrols/fusion/impl/TextFieldBackground.qml
new file mode 100644
index 0000000000..5626d9b3dd
--- /dev/null
+++ b/src/quickcontrols/fusion/impl/TextFieldBackground.qml
@@ -0,0 +1,35 @@
+// Copyright (C) 2025 The Qt Company Ltd.
+// SPDX-License-Identifier: LicenseRef-Qt-Commercial OR LGPL-3.0-only OR GPL-2.0-only OR GPL-3.0-only
+
+import QtQuick
+import QtQuick.Controls.impl
+import QtQuick.Controls.Fusion
+
+Rectangle {
+    implicitWidth: 120
+    implicitHeight: 24
+    radius: 2
+    color: control.palette.base
+    border.color: control.activeFocus ? Fusion.highlightedOutline(control.palette) : Fusion.outline(control.palette)
+
+    required property Item control
+
+    Rectangle {
+        x: 1
+        y: 1
+        width: parent.width - 2
+        height: parent.height - 2
+        color: "transparent"
+        border.color: Color.transparent(Fusion.highlightedOutline(control.palette), 40 / 255)
+        visible: control.activeFocus
+        radius: 1.7
+    }
+
+    Rectangle {
+        x: 2
+        y: 1
+        width: parent.width - 4
+        height: 1
+        color: Fusion.topShadow
+    }
+}
-- 
2.51.2

